src/nomad_simulations/physical_property.py

```diff
@@ -0,0 +1,231 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Any, Optional
+
+from nomad import utils
+from nomad.datamodel.data import ArchiveSection
+from nomad.metainfo import (
+    Quantity,
+    SubSection,
+    SectionProxy,
+    Reference,
+    Section,
+    Context,
+    MEnum,
+)
+from nomad.metainfo.metainfo import DirectQuantity, Dimension, _placeholder_quantity
+from nomad.datamodel.metainfo.basesections import Entity
+
+from .variables import Variables
+from .numerical_settings import SelfConsistency
+
+
+# We add `logger` for the `PhysicalProperty.variables_shape` method
+logger = utils.get_logger(__name__)
+
+
+class PhysicalProperty(ArchiveSection):
+    """
+    A base section used to define the physical properties obtained in a simulation, experiment, or in a post-processing
+    analysis. Besides its `value`, it also contains:
+        - Quantities that are string identifiers: `source`, `type`, `label`
+        - Quantities for calculating the shape of the physical property: `shape`, `variables`
+        - Quantities for referencing and establishing the character of a physical property: `outputs_ref`, `entity_ref`, `self_consistent_ref`,
+        `is_derived`, `is_scf_converged`
+    """
+
+    name = Quantity(
+        type=str,
+        description="""
+        Name of the physical property. Example: `'ElectronicBandGap'`.
+        """,
+    )
+
+    source = Quantity(
+        type=MEnum('simulation', 'measurement', 'analysis'),
+        default='simulation',
+        description="""
+        Source of the physical property. Example: an `ElectronicBandGap` can be obtained from a `'simulation'`
+        or in an `'measurement'`.
+        """,
+    )
+
+    type = Quantity(
+        type=str,
+        description="""
+        Type categorization of the physical property. Example: an `ElectronicBandGap` can be `'direct'`
+        or `'indirect'`.
+        """,
+        # ! add more examples in the description
+    )
+
+    label = Quantity(
+        type=str,
+        description="""
+        Label for additional classification of the physical property. Example: an `ElectronicBandGap`
+        can be labeled as `'DFT'` or `'GW'` depending on the methodology used to calculate it.
+        """,
+        # ! add more examples in the description
+    )
+
+    rank = DirectQuantity(
```

Comment:
Is the use of `DirectQuantity` instead of `Quantity` necessary here? The docs specify that it resolves _bootstrapping_.
Or is this just a hack to circumvent the typing system of `Quantity`?

Reply:
When I was using `Quantity`, I could not manipulate lists to concatenate them so that `full_shape` is a Dimension as well and can be passed to the `value.shape`. I can ask Theodore about this.

Reply:
So, for your second question: yes